### PR TITLE
check-aux: test core, alloc, std in Miri

### PR DESCRIPTION
**Pull request by RalfJung**
Let's see if this works, and how long it takes.

### CODE REVIEW

**Comment by RalfJung** [comment]
Oh...
```
     Running benches/lib.rs (build/x86_64-unknown-linux-gnu/stage1-std/miri/x86_64-unknown-linux-gnu/debug/deps/allocbenches-a59974b85394f408)
```
This runs some benchmarks it seems. That will ~never terminate under Miri.^^

**Comment by RalfJung** [comment]
Oh, interesting... the macOS build of compiler-builtins is trying to run a C compiler.

For now let's just run tests on Linux only, then.

**Comment by RalfJung** [comment]
Strangely the macos tests work fine locally. Not sure what is different on CI...

**Comment by RalfJung** [comment]
All right, tests are passing!
- core, alloc lib tests: 20 minutes
- core, alloc doc tests: 20 minutes
- std lib tests: 3 minutes
- std doc tests: 1 minute

On std we're unfortunately only testing one target as somehow the compiler-builtins build script fails for other targets. I'll extend the std tests a bit (we need some filters here as not all of std is supported by Miri).

There are likely some testcases that are unproportionally slow. They are somewhat hard to identify. For doctests we get the "has been running for a long time" warning, which points at [this test](https://doc.rust-lang.org/nightly/std/sync/struct.Arc.html#method.into_inner) that creates a 100k element linked list. For libtests, the test harness runs inside Miri where it disables timing in Miri to make sure it can run with isolation enabled. Now that we have the fake clock in isolation mode, maybe we change fix that.

File: library/alloc/src/sync.rs [hunk 1]
```diff
@@ -1062,7 +1062,9 @@ impl<T, A: Allocator> Arc<T, A> {
     ///
     /// // Create a long list and clone it
     /// let mut x = LinkedList::new();
-    /// for i in 0..100000 {
+    /// let size = 100000;
+    /// # let size = if cfg!(miri) { 100 } else { size };
+    /// for i in 0..size {
```
Comment (RalfJung) [inline library/alloc/src/sync.rs:1067]: Is that an okay approach to make the test shorter in Miri? The alternative would be more convoluted but would let the code look as before on the web view:
```rust
# let size = if cfg!(miri) { 100 } else { 100000 };
# for i in 0..size {
# /*
for i in 0..100000 {
# */
```

**Comment by RalfJung** [comment]
It's actually a bit surprising that these take so long, since the out-of-tree test tests alloc and core in 45 minutes total and it runs all tests on 2 targets. So the test here should have been about twice as fast.

Do we have debug assertions enabled on the gnu-aux runner, or something like that?

**Comment by RalfJung** [comment]
Cc @Mark-Simulacrum

**Comment by ChrisDenton** [comment]
> Oh, interesting... the [macOS build of compiler-builtins is trying to run a C compiler](https://github.com/rust-lang/rust/actions/runs/8579217803/job/23513955276).

presumably the `build.optimized-compiler-builtins` config is set to `true`.

**Comment by Mark-Simulacrum** [comment]
> > Oh, interesting... the [macOS build of compiler-builtins is trying to run a C compiler](https://github.com/rust-lang/rust/actions/runs/8579217803/job/23513955276).
> 
> presumably the `build.optimized-compiler-builtins` config is set to `true`.

Yes: https://github.com/rust-lang/rust/blob/01f7f3a1ff983ddfe485bbd9cedd59c378fdaceb/src/ci/run.sh#L91

> Do we have debug assertions enabled on the gnu-aux runner, or something like that?

Yes: https://github.com/rust-lang/rust/blob/01f7f3a1ff983ddfe485bbd9cedd59c378fdaceb/src/ci/run.sh#L125-L140

**Comment by RalfJung** [comment]
> presumably the build.optimized-compiler-builtins config is set to true.

Good catch, thanks! I am trying with std tests now. Not sure if they will all work on all targets though, we will see.

Regarding debug assertions, that would probably explain it. 2x sounds like a plausible overhead for that.

**Comment by RalfJung** [comment]
[With debug assertions but without overflow checks:](https://github.com/rust-lang/rust/actions/runs/8582813293/job/23521327539?pr=123506)
- core+alloc lib tests: 21min
- core+alloc doc tests: 17min
- std lib tests: 12min
- std doc tests: 2min
- some std lib tests on other targets: 16min

Looks like that's not a lot faster than with overflow checks. So it's the debug assertions that are expensive, not the overflow checks.

**Comment by RalfJung** [comment]
[With overflow checks but without debug assertions:](https://github.com/rust-lang/rust/actions/runs/8583713400/job/23523236247?pr=123506)
-   core+alloc lib tests: 20min
-   core+alloc doc tests: 17min
-   std lib tests: 12min
-   std doc tests: 2min
-   some std lib tests on other targets: 15min

Either the assertions make no difference or, more likely, I am setting these environment variables in the wrong place...

File: src/bootstrap/mk/Makefile.in [hunk 2]
```diff
@@ -52,6 +52,36 @@ check-aux:
...
+	# Also, they work only on the host.
+	$(Q)MIRIFLAGS="-Zmiri-ignore-leaks -Zmiri-disable-isolation" \
+		$(BOOTSTRAP) miri --stage 2 \
+		library/core \
+		library/alloc \
+		--doc
+	# In `std` we cannot test everything.
+	$(Q)MIRIFLAGS="-Zmiri-disable-isolation" BOOTSTRAP_SKIP_TARGET_SANITY=1 \
+		$(BOOTSTRAP) miri --stage 2 library/std \
+		--no-doc -- \
```
Comment (RalfJung) [inline src/bootstrap/mk/Makefile.in:73]: I didn't add mips-unknown-linux-gnu here as I think std has much less code that depends on target or endianess, so the extra 10min CI time does not seem worth it.

**Comment by RalfJung** [comment]
Ah, that's more like it. Now even when [testing two targets for core and alloc](https://github.com/rust-lang/rust/actions/runs/8586490306/job/23529017526?pr=123560) we get
- core+alloc libtests: 30min
- core+alloc doctests: 10min
- std libtests: 10min
- std doctests: 1min
- std reduced libtests for extra targets: 11min

The total duration of the gnu-aux job is around 104min.

That is just above the anticipated 1h. I can remove some of the extra targets if that is desired. Testing core and alloc only on one target brings it down to 90min.

r? @Mark-Simulacrum

**Comment by RalfJung** [comment]
Some data from https://github.com/rust-lang/rust/pull/123560: enabling just overflow checks add around 10min to total CI time, enabling just debug assertions adds around 30min.

File: library/alloc/src/sync.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply library/alloc/src/sync.rs:1067]: Seems fine to me, especially given it's relatively isolated. Realistically I'm not convinced this example needs a huge length (100 elements or 100,000 is about equally bad if you're pushing a new stack frame per element, IMO, given that you might already be close to the end of the allowed stack when starting to execute).

**Comment by Mark-Simulacrum** [comment]
@bors r+

I think this is good to land. We can always tweak it further in future PRs (e.g., get rid of the Makefile, maybe try to enable benchmarks running once or something like that).

**Comment by RalfJung** [comment]
I think benchmarks are already only run once in test mode? This is based on a vague memory, nothing more.

However some benchmarks create large data structures in the preparation phase of the benchmark and I think that is what makes them slow on Miri.

**Comment by Kobzol** [comment]
This PR had nothing to do with our benchmarking infrastructure :) So this has to be pure noise.

**Pull request merged**